design/0005-circuit-breaking.md

```diff
@@ -0,0 +1,150 @@
+all the proxies i.e. we **won't** provide user ability to configure it just for
+proxy which raised those failures due to which ejection got triggered.
+
+Let me explain above piece with an example. Let's say there are two services A
+and B which calls C service. In case B->C gets success rate of 100% but A->C
+gets success rate of 50%, it won't be possible for user to configure
+service-mesh such that it would trigger circuit breaking only for A->C traffic.
+Instead either circuit breaking will happen for both A and B (through
+aggregation of their stats) or none of them.
+
+This certainly has its own disadvantages but I would be somewhat inclined
+towards having it simpler unless usecases for making it per-proxy over-weighs
+other one. So, for now, I propose that every ejection policy would be based on
+taking aggregated metrics over all the proxy instances to mark any instance for
+ejection.
+
+For HTTP/gRPC, I feel like taking advantage of linkerd provided golden metrics
+for ejection policies. Ejection can be based on success-rate,
+percentile-latencies for a given time window. Circuit breaker settings would be
+applied per service + protocol instead of per-route. In case circuit breaker is
+triggered for entire service, then we will return 503 for HTTP with some custom
+header and gRPC message for gRPC.
+
+If we go with the above design, may be we can keep most of the ejection logic in
+control plane i.e. in the destination API, we can just remove ejected endpoints
+from list given to discovery.
+
+Other than above mentioned details, here's the top level changes I could think
+of with this approach -
+
+1. Updating ServiceProfile CRD and its validation.
+2. Updating destination service to use circuit breaking configurations specified
+   by user to perform ejection. It can use prometheus metrics for evaluation and
+   update endpoints as required.
+3. Sharing some information through destination API to linkerd2-proxy so that it
+   can return adequate response to the requests.
+4. AFAIK metrics exposed by each proxy doesn't container upstream service
+   instance's endpoint.
+5. Exposing metrics related to circuit breaking from control plane.
+
+There's lot more things which needs to be discussed but I would first like to
+get some feedback if basic idea sounds sane to everyone.
+
+### Prior art
+
+[prior-art]: #prior-art
+
+Other service meshes seems to be performing this in data plane instead, may be
+because it aligned well with their architecture. All the service meshes built
+over top of enovy, uses envoy's inbuilt capability of circuit breaking. In
+[envoy](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier),
+they use two terms - "Circuit Breaking" and "Outlier Detection", which I feel
+seems inter-related upto some extent and can be merged into single. Envoy
+supports both cluster-wide as well as per-proxy outlier detectors.
```

Comment:
AFAIK "cluster-wide" in Envoy parlance just means a configuration group right? You'd need to implement an xDS controller that consumes per-envoy metrics to orchestrate multiple envoy proxies?